SLING-7792 - Resource Resolver should return more than one resolved path if available

rombert opened the pull request:
API draft for the new `ResourceMapper` service.

Highlights (all up for debate, obviously):

- deprecated all `map()` methods from the `ResourceResolver` and pointed to the `ResourceMapper`
- no longer have methods which only take a `String` argument. Now all methods take a `String` and a `HttpServletRequest`. I did this since I think two methods are better than four :-)
- argument order is reversed - since the `HttpServletRequest` is optional, it's the last one
- I expect the `ResourceMapper` implementation to be based on a `ResourceResolver` and the javadoc reflects that
- rather than keeping the existing `map` method and adding a `maps` one, I named then `getMapping` and `getAllMappings`. @bdelacretaz proposed `getMappingCandidates` but I am not sure if `getMappingCandidate` would be a good replacement for the `map` method, so I went with `getMapping`

JEBailey commented:
What's the reasoning for having a parameter for HttpServletRequest at all?

rombert commented:
@JEBailey - in that case a full URL is returned, not just the repository path.

rombert commented:
@JEBailey - I've updated the PR, in case you want to add more comments

cziegeler commented:
I would leave the javadocs in ResourceResolver for those methods - otherwise you don't know what the parameters are, what type of exception might be thrown etc.
Instead of the import statement for the mapper in resource resolver, add the full qualified class to the link. This way we don't have a code dependency from the resource package to the new package.

rombert commented:
@cziegeler  - I updated the Javadoc . Sorry for not replying regarding that, I did not see that package used use due to the javadoc links so I left those with imports:

```
org.apache.sling.api.resource.mapping;version="1.0.0";uses:="javax.annotation,javax.servlet.http"
```

But it's fine either way so I updated the javadocs as you suggested in the latest commit.

rombert commented:
I think we have consensus here, so merging.

rombert commented:
... and I messed up the ResourceResolver javadoc while squashing. WIll fix with a follow-up commit.

pun-ky commented:
@rombert  could you take into consideration also implementing this - https://issues.apache.org/jira/browse/SLING-4327 ? Still it is unaddressed problem on AEM. Regardless implementation details, now it is not possible to generate on author instance a mapped link pointing to publish instance (when resource resolver mappings root path is different for each run mode). 

This case is needed, for example, if project wants to implement content moderation being performed on author which ends with sending e-mail to user-generated content authors informing them that their content has been approved / assuming the link is being a part of that e-mail. That was a real business project scenario and as I remember there was few others, because of that limitation in Sling.

Maybe we could do something with that together? Kind regards, Krystian.
